src/Switch.js

```diff
@@ -1,59 +1,64 @@
-import React, { PropTypes } from 'react';
-import TabBar from './TabBar';
-import { assert } from './Util';
-
-export default function Switch(props) {
-  const navState = props.navigationState;
-  const selector = props.selector;
-  const statem = props.statem;
-  if (!selector && !statem) console.error('Selector should be defined.');
-  let index = -1;
-  if (!selector) {
-    // support Statem - Harel statecharts machine!
-    navState.children.forEach((el, i) => {
-      assert(el.default || el.state,
-        `Either default or state should be defined for element=${el.key}`);
-      if (el.default) {
-        index = i;
-      } else {
-        if (el.state.active) {
-          index = i;
-        }
-      }
-    });
-    assert(index !== -1, 'No default scene is defined');
-  } else {
+import React, { Component, PropTypes } from 'react';
+
+import DefaultRenderer from './DefaultRenderer';
+import Actions from './Actions';
+
+export default class Switch extends Component {
+
+  static propTypes = {
+    navigationState: PropTypes.object,
+    onNavigate: PropTypes.func,
+    selector: PropTypes.func,
+    unmountScenes: PropTypes.bool,
+  };
+
+  constructor(props, context) {
+    super(props, context);
+    this.updateState = this.updateState.bind(this);
+    this.state = {};
+  }
+
+  componentDidMount() {
+    this.updateState(this.props);
+  }
+
+  componentWillReceiveProps(props) {
+    this.updateState(props);
+  }
+
+  updateState(props) {
+    const navState = props.navigationState;
+
+    const selector = props.selector;
+    if (!selector) console.error('Selector should be defined.');
+
     const selectedKey = selector(props);
     if (!selectedKey) console.error('Selector should return key.');
-    navState.children.forEach((el, i) => {
-      if (el.sceneKey === selectedKey) {
-        index = i;
+
+    const selected = navState.children.filter(el => el.sceneKey === selectedKey);
+    if (!selected) console.error(`A scene for key “${selectedKey}” does not exist.`);
+
+    const navigationState = selected[0];
+    if (!navigationState) console.error(`Cannot find a scene with key “${selectedKey}”`);
+
+    if (navigationState.key !== navState.children[navState.index].key) {
+      if (props.unmountScenes) {
+        Actions[selectedKey]({ unmountScenes: true });
       }
-    });
-    if (index === -1) console.error(`A scene for key “${selectedKey}” does not exist.`);
+      Actions[selectedKey]();
+    }
+    this.setState({ navigationState });
```

Comment:
it will do `this.setState()` in the end using variable `navigationState` ( a child of `navState` )
so if the **second** time you are entering `this.updateState()`, navState would never match scene key returned by selector.

the first time you entered `this.updateState()`:  `navState.key` is 'switch', and it's children are ['login', 'home']
the second time you entered `this.updateState()`:  `navState.key` is 'home', and it's children are ['tab1', 'tab2']

my question: is `navState` supposed to be the child one eventually ?